---
title: "Quantifying the overlap of ASVs between taxa"
author: adimascf
date: "2022-12-16"
output:
    github_document:
        html_preview: false
editor_options: 
  chunk_output_type: inline
---


```{r, message=FALSE}
library(tidyverse)
library(here)
library(knitr)

metadata <- read_tsv(here("data/references/genome_id_taxonomy.tsv"),
                     col_types = cols(.default = col_character())) %>%
    mutate(strain = if_else(scientific_name == species, 
                            NA_character_, scientific_name)) %>%
    select(-scientific_name) 

asv <- read_tsv(here("data/processed/rrnDB.count_tibble"),
                col_types = cols(.default = col_character(),
                                 count = col_integer()))

metadata_asv <- inner_join(metadata, asv, by=c("genome_id" = "genome"))
```


### How often is the same ASV found in multiple taca from the same rank?

Across the analyses, we have already seen that a single bacterial genome can have multiple ASVs and that depending on the species we are looking at, there may be as many ASVs as there are genome sequences for that species. To me, this says that ASVs pose a risk at unnaturally splitting a species and even a genome into multiplt taxonomic groupings! In my book, that is not good.


Today, I want to ask an equally important question: If I have an ASV, what is the probability that it is also found in another taxonomic group from the same rank? In other words, if I have an ASV from Bacillus subtilis, what is the probability of finding it in Bacillus cereus? Of course, it is probably more likely to find a Bacillus subtillis ASV in a more closely related organism like Bacillus cereus than E. coli. Perhaps we can control for relatedness in a future episode, but for today, I want to answer this question for any two taxa from the same rank.


```{r}

# metadata_asv - input data
overlap_data <- metadata_asv %>%
# - focus on taxonomic ranks - kingdom to species, asv, and region
    select(-genome_id, -count, -strain) %>%
# - make data frame tidy
    pivot_longer(cols = c(-region, -asv), names_to = "rank", 
                 values_to = "taxon") %>%
# - remove line from the data where we do not a taxonomy
    drop_na(taxon) %>%
# - remove redundant lines
    distinct() %>%

# for each region and taxonomic rank, group by asvs `
    group_by(region, rank, asv) %>%
# - for each asv - count the number of taxa
    summarize(n_taxa=n(), .groups = "drop_last") %>%
# - count the number of asvs that appear in more than one taxa
    count(n_taxa) %>%
# - find the ratio od the number of asvs in more than one taxa to the total number of asvs
    summarize(overlap = sum((n_taxa > 1) * n) / sum(n), .groups = "drop") %>%
    mutate(rank = factor(rank,
                         levels = c("kingdom", "phylum", "class",
                                    "order", "family", "genus", "species"))) %>%
    mutate(overlap = 100*overlap)


# Create a plot showing specificity at each taxonomic rank for each region
# x = taxonomic rank
# y = specificity or % of asvs found in more than one taxa
# geom = line plot
# different lines for each region of 16S rRNA gene
overlap_data %>%
    ggplot(aes(x = rank, y = overlap, group = region, color=region)) +
    geom_line()

overlap_data %>%
    filter(rank == "species") %>%
    kable()



```


### Conclusion
* Sub region are less specific at the species level than full-length. Still full-length has 3.6% overlap whereas the subregion are between 8.7 and 12.5% overlap
* This analysis does not control for uneven sampling of different species

